neuro_san/internals/run_context/langchain/core/langchain_run_context.py

```diff
@@ -571,171 +331,27 @@ async def wait_on_run(self, run: Run, journal: Journal = None) -> Run:
             "input": self.recent_human_message.content
         }
 
-        # Create the list of callbacks to pass when invoking
-        parent_origin: List[Dict[str, Any]] = self.get_origin()
-        base_journal: Journal = self.invocation_context.get_journal()
-        origination: Origination = self.invocation_context.get_origination()
-        callbacks: List[BaseCallbackHandler] = [
-            JournalingCallbackHandler(self.journal, base_journal, parent_origin, origination)
-        ]
-        # Consult the agent spec for level of verbosity as it pertains to callbacks.
-        agent_spec: Dict[str, Any] = self.tool_caller.get_agent_tool_spec()
-        verbose: Union[bool, str] = agent_spec.get("verbose", False)
-        if isinstance(verbose, str) and verbose.lower() in ("extra", "logging"):
-            # This particular class adds a *lot* of very detailed messages
-            # to the logs.  Add this because some people are interested in it.
-            callbacks.append(LoggingCallbackHandler(self.logger))
-
-        # Set up a run name for tracing purposes
-        metadata: Dict[str, Any] = self.invocation_context.get_metadata()
-        run_name: str = metadata.get("request_id", "<unknown>") + "-" + \
-            Origination.get_full_name_from_origin(self.origin)
-
-        # Add callbacks as an invoke config
-        invoke_config = {
-            "configurable": {
-                "session_id": run.get_id()
-            },
-            "callbacks": callbacks,
-            "recursion_limit": recursion_limit,
-            "run_name": run_name
-        }
+        run: Run = LangChainRun(self.run_id_base, self.chat_history)
+        session_id: str = run.get_id()
 
-        # Chat history is updated in write_message
-        await self.journal.write_message(self.recent_human_message)
+        runnable = NeuroSanRunnable(agent_chain=self.agent_chain,
+                                    primary_llm=self.llm_resources.get_model(),
+                                    invocation_context=self.invocation_context,
+                                    journal=self.journal,
+                                    interceptor=self.interceptor,
+                                    origin=self.origin,
+                                    tool_caller=self.tool_caller,
+                                    error_detector=self.error_detector,
+                                    session_id=session_id)
+        runnable_config: Dict[str, Any] = runnable.prepare_runnable_config(session_id=session_id, use_run_name=True)
```

Comment:
Much lighter weight method implementation now as the functionality that starts here gets delegated to NeuroSanRunnable